oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/Optimizer.kt

```diff
@@ -0,0 +1,358 @@
+                    OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(name)
+            ).build()
+
+            sim.input.onNext(message)
+        }
+    }
+
+    fun cancel(name: String) = GlobalScope.launch {
+        val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(name)
+        ).build()
+
+        simulationsByName.getValue(name).input.onNext(message)
+    }
+
+    fun setDuration(nodeName: String?, timeOut: Duration?) {
+        if (nodeName == null || !simulationsByName.containsKey(nodeName)) {
+            return
+        }
+        val node: Simulation = simulationsByName.getValue(nodeName)
+        val newNode = node.copy(
+                timeOut = timeOut
+        )
+        simulationsByName += nodeName to newNode
+    }
+
+    fun disconnectAll() = GlobalScope.launch {
+        for ((name, sim) in simulationsByName) {
+            try {
+                sim.input.onCompleted()
+            } catch (e: StatusRuntimeException) {
+                println("Error when close input for:\n$e")
+            }
+        }
+    }
+
+    fun cancelAndStop() {
+        GlobalScope.launch {
+            stopOptimization()
+            for ((name, sim) in simulationsByName) {
+                val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                        OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(name)
+                ).build()
+                sim.input.onNext(message)
+            }
+        }
+    }
+
+    private suspend fun syncConfigFor(simName: String) {
+        val sim = simulationsByName.getValue(simName)
+        val message = OASISQueryDTO.newBuilder().setNodeStatusRequest(
+                OASISQueryDTO.NodeStatusUpdateRequest.newBuilder().setName(simName)
+        ).build()
+
+        sim.input.onNext(message)
+
+        val result = select<Either<Simulation, Message>> {
+            sim.update.onReceive { it -> Either.Left(updateFromResponse(it)) }
+            sim.error.onReceive { it ->
+                Either.Right(
+                        Message(LocalDateTime.now(), it.name, "Error update node $simName due to ${it.message} :\n${it.exception}")
+                )
+            }
+        }
+
+        if (result.isLeft()) {
+            simulationsByName += simName to result.left().get()
+        } else {
+            viewData.addMessage(result.right().get())
+        }
+    }
+
+
+    override fun stopOptimization(request: StopOptimizationCommandDTO, responseObserver: StreamObserver<StopOptimizationResponseDTO>) = responseObserver.consume {
+        stopOptimization()
+        StopOptimizationResponseDTO.newBuilder().setMessage("Stopped").setStopped(true).build()
+    }
+
+
+    override fun offerSimulationResult(request: SimulationResponseDTO, responseObserver: StreamObserver<SimulationResultConfirmDTO>) = responseObserver.consume {
```

Comment:
ahh I see, so this object is both charged with converting from the GRPC method sources to channels, and then tasked with decoding the purpose of those channels.

Sounds like two objects, doesnt it?